### PR TITLE
feat: sort txs by indirect dep

### PR DESCRIPTION
**Pull request by zhangsoledad**
Previously,  BlockAssembler selection has not considered such a situation

[](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggTFJcbiAgIHR4MSAtLmRlcC4tPiB0eDIgXG4gICB0eDEgLS1pbnB1dC0tPiB0eDMgXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

This pr rewrite this part, make sure transaction selection including all unconfirmed ancestors.
[](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggTFJcbiAgIHR4MSAtLmRlcC4tPiB0eDIgXG4gICB0eDEgLS1pbnB1dC0tPiB0eDMgXG4gICB0eDIgLS4uLi0gdHgzIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

### CODE REVIEW

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 445
  - CKB Version: ee8011908d1b896db33d2f8c48ed9483f4fde453
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 441
  - CKB Version: ee8011908d1b896db33d2f8c48ed9483f4fde453
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 479
  - CKB Version: 1b88ea0876f19a629a1bbe8bb4cf4455b4375c4d
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 475
  - CKB Version: 63fa58e343152b8e1922ba0b0c5471e9f785d93e
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by nervos-bot-user** [comment]
Can one of the admins verify this patch?

**Comment by doitian** [comment]
In this PR, if the pool already accepted tx3, will it accept tx2?

**Comment by zhangsoledad** [comment]
> In this PR, if the pool already accepted tx3, will it accept tx2?

if tx3 accepted before tx2, tx2 will be discarded, I add a test case describe this.

**Review by driftluo — DISMISSED** [review]

**Review by quake — APPROVED** [review]

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 477
  - CKB Version: 166c8b7dc3929b3d2251f98f10937a9015abb91c
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Review by driftluo — APPROVED** [review]

**Comment by driftluo** [comment]
bors r=driftluo,quake

**Comment by zhangsoledad** [comment]
bors try

**Comment by zhangsoledad** [comment]
bors try

**Review by quake — DISMISSED** [review]

File: test/src/specs/tx_pool/different_txs_with_same_input.rs [hunk 1]
```diff
@@ -47,11 +44,12 @@ impl Spec for DifferentTxsWithSameInput {
             .map(TransactionView::hash)
             .collect();
 
-        assert!(commit_txs_hash.contains(&tx2.hash()));
-        assert!(!commit_txs_hash.contains(&tx1.hash()));
+        // RBF (Replace-By-Fees) is not implemented
+        assert!(commit_txs_hash.contains(&tx1.hash()));
+        assert!(!commit_txs_hash.contains(&tx2.hash()));
 
         // when tx2 was confirmed, tx1 should be discarded
```
Comment (driftluo) [inline test/src/specs/tx_pool/different_txs_with_same_input.rs:51]: The comment is wrong

Comment (zhangsoledad) [reply test/src/specs/tx_pool/different_txs_with_same_input.rs:51]: I don't get it. what's wrong.

**Review by quake — APPROVED** [review]

**Review by driftluo — APPROVED** [review]

**Comment by zhangsoledad** [comment]
bors r=quake,driftluo

**Pull request merged**